logstash-core/lib/logstash/api/commands/stats.rb

```diff
@@ -121,8 +146,17 @@ def report(stats)
               },
               :reloads => stats[:reloads],
               :queue => stats[:queue]
-            }.merge(stats[:dlq] ? {:dead_letter_queue => stats[:dlq]} : {})
+            }
+            ret[:dead_letter_queue] = stats[:dlq] if stats.include?(:dlq)
+
+            # if extended_stats were provided, enrich the return value
+            if extended_stats
+              ret[:queue]    = extended_stats["queue"] if extended_stats.include?("queue")
+              ret[:vertices] = extended_stats["vertices"] if opts[:vertices] && extended_stats.include?("vertices")
```

Comment:
I think this might be the ideal spot to inject the decoration of vertices with plugin metadata.

```suggestion
              if opts[:vertices] && extended_stats.include?("vertices")
                ret[:vertices] = extended_stats["vertices"].map { |vertex| decorate_vertex(vertex) }
              end
```

Which would work with the following private method. Note that we explicitly avoid mutating the vertices entries, as these are shared memory structures that outlive our individual API requests.

~~~ ruby
          private
          ##
          # Returns a vertex, decorated with additional metadata if available.
          # Does not mutate the passed `vertex` object.
          #
          # @api private
          # @param vertex [Hash{String=>Object}]
          # @return [Hash{String=>Object}]
          def decorate_vertex(vertex)
            plugin_id = vertex["id"]&.to_s
            return vertex unless plugin_id && LogStash::PluginMetadata.exists?(plugin_id)

            plugin_metadata = LogStash::PluginMetadata.for_plugin(plugin_id)
            cluster_uuid = plugin_metadata&.get(:cluster_uuid)
            vertex = vertex.merge("cluster_uuid" => cluster_uuid) unless cluster_uuid.nil?

            vertex
          end
~~~

Reply:
Hi @yaauie 

This does not seem to work for me when the `decorate_vertex` method is marked as `private`. 

Here is the changed section of the file:

```
        module PluginsStats
          module_function

          def plugin_stats(stats, plugin_type)
            # Turn the `plugins` stats hash into an array of [ {}, {}, ... ]
            # This is to produce an array of data points, one point for each
            # plugin instance.
            return [] unless stats[:plugins] && stats[:plugins].include?(plugin_type)
            stats[:plugins][plugin_type].collect do |id, data|
              { :id => id }.merge(data)
            end
          end

          private
          module_function
          ##
          # Returns a vertex, decorated with additional metadata if available.
          # Does not mutate the passed `vertex` object.
          #
          # @api private
          # @param vertex [Hash{String=>Object}]
          # @return [Hash{String=>Object}]
          def decorate_vertex(vertex)
            plugin_id = vertex["id"]&.to_s
            return vertex unless plugin_id && LogStash::PluginMetadata.exists?(plugin_id)

            plugin_metadata = LogStash::PluginMetadata.for_plugin(plugin_id)
            cluster_uuid = plugin_metadata&.get(:cluster_uuid)
            vertex = vertex.merge("cluster_uuid" => cluster_uuid) unless cluster_uuid.nil?

            vertex
          end

          def report(stats, extended_stats=nil, opts={})
            ret = {
              :events => stats[:events],
              :plugins => {
                :inputs => plugin_stats(stats, :inputs),
                :codecs => plugin_stats(stats, :codecs),
                :filters => plugin_stats(stats, :filters),
                :outputs => plugin_stats(stats, :outputs)
              },
              :reloads => stats[:reloads],
              :queue => stats[:queue]
            }
            ret[:dead_letter_queue] = stats[:dlq] if stats.include?(:dlq)

            # if extended_stats were provided, enrich the return value
            if extended_stats
              ret[:queue]    = extended_stats["queue"] if extended_stats.include?("queue")
              if opts[:vertices] && extended_stats.include?("vertices")
                ret[:vertices] = extended_stats["vertices"].map { |vertex| decorate_vertex(vertex) }
              end
            end

<snip>
```

It does work correctly when marking `decorate_vertex` as a `private_class_method` instead of simply `private`, but I'm afraid I don't know if this achieves the effect you were intending.

The good news is that this does work correctly when the `private` marker is removed.

Reply:
ah. I missed the `Module#module_function` declaration at the top of the module (which makes the subsequently-defined instance methods private and also extends them publicly onto the module itself), so we can just avoid sending `Module#private` entirely.

***

Since `PluginStats#decorate_vertex` is an _implementation detail_ defined to make `PluginStats#report` more readable, it is effectively a private-API method and should be declared at the _bottom_ of this module definition.

While you may see plenty of evidence to the contrary in this project, by convention in Ruby, we tend to declare methods in an order that indicates publicity instead of oscillating between modes: first all of the public methods, then protected, then private. This allows a reader to quickly understand a module or class's public API quickly, to be aware of the scope of effect of a change to a class's implementation, and to avoid tying one class's behaviour to another class's implementation details.

Reply:
Hi @yaauie. I've removed the `private` declaration and I believe that's the last outstanding issue on this PR. Thanks!